### PR TITLE
Smsalert

### PR DESCRIPTION
**Pull request by pamzey**
I have added sms controller and service to be used for notifications

### CODE REVIEW

File: config/initializers/sidekiq.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# config/initializers/sidekiq.rb
+Sidekiq.default_worker_options = { 'retry' => true } # Retry indefinitely
+
```
Comment (fchiyenda) [inline config/initializers/sidekiq.rb:3]: Line crushing when trying to run the application
```suggestion
Sidekiq.default_job_options = { 'retry' => true } # Retry indefinitely
```

File: app/controllers/api/v1/immunization_reports_controller.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        else
+          render status: :no_content
+        end
+      end
+
+
+      private
+
+      def validate_params
+        permitted = params.permit(%i[start_date end_date name quarter year]).to_h
```
Comment (fchiyenda) [inline app/controllers/api/v1/immunization_reports_controller.rb]: I think we should also handle required params

File: app/services/immunization_service/send_sms_service.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      @details = details
+      @converted_date = @date.strftime('%d-%B-%Y')
+            
+      message = "Okondweda #{@details[:firstname]} #{@details[:sirname]},\n" \
+                "Mukukuziwitsidwa kuzabwelaso ku Chiptala\n" \
+                "pa tsiku **#{@converted_date}**.\n" \
+                "kut muzalandileso katemela.\n" \
+                "Zikomo."
+
+      uri = URI.parse("https://gateway.seven.io/api/sms")
```
Comment (fchiyenda) [inline app/services/immunization_service/send_sms_service.rb]: I am thinking API key and gateway should be in a configuration file. eg application.yml

**Review by fchiyenda — CHANGES_REQUESTED** [review]

File: app/services/immunization_service/send_sms_service.rb (hunk 3 above)
Comment (pamzey) [reply app/services/immunization_service/send_sms_service.rb]: I have added all the suggested changes tried it and the job logs seems to OK

File: config/initializers/load_application_yml.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# config/initializers/load_application_yml.rb
+config = YAML.load_file(Rails.root.join('config', 'application.yml'))[Rails.env]
```
Comment (fchiyenda) [inline config/initializers/load_application_yml.rb]: Let us ensure YAML aliases are enabled usually it breaks the code:
```suggestion
config = YAML.safe_load(File.read(Rails.root.join('config', 'application.yml')), aliases: true)[Rails.env]
```

File: config/application.yml.example [hunk 5]
```diff
@@ -1,104 +0,0 @@
-# Comment out parameter below to enable syncing with lims through the REST API.
```
Comment (fchiyenda) [inline config/application.yml.example]: Chief I think you accidentally deleted application.yml.example

Comment (pamzey) [reply config/application.yml.example]: i have replaced the file and added the suggestion

File: app/jobs/send_sms_job.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        "to" => @details[:cell_phone],
+        "from" => "IMMUNIZATION APP",
+        "text" => message
+      )
+  
+      response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) do |http|
+        http.request(request)
+      end
+
+      Rails.logger.info("SMS Gateway Response........pamzey: #{response.code} - #{response.body}")
```
Comment (fchiyenda) [inline app/jobs/send_sms_job.rb]: Let us remove the personalization

Comment (pamzey) [reply app/jobs/send_sms_job.rb]: i have removed the error logger

Comment (fchiyenda) [reply app/jobs/send_sms_job.rb]: Meant removing your name in the code

File: config/application.yml.example (hunk 5 above)
Comment (pamzey) [reply config/application.yml.example]: i have added the application.yml.example file

**Review by fchiyenda — CHANGES_REQUESTED** [review]
Let us not push the key to github

**Comment by pamzey** [comment]
I have added guide to the read me how to get the sms api key

**Review by fchiyenda — COMMENTED** [review]
> i have added the application.yml.example file 

I think you did not replace the first line which is informative for people who are setting up.

**Comment by pamzey** [comment]
I have removed the line. The reason I left it there was to inform the user about where we purchased our SMS credit, in case they are new. However, I understand the security issue, so I have removed it.

File: README.md [hunk 7]
```diff
@@ -407,3 +407,10 @@ See the following for more:
 * You don't need to worry about setting up the gems, the container comes with them pre-installed
 
 [Back to top](#table-of-contents)
+
+# SMS Gateway for the immunization app
+  * You need to ask any of the team members for the api key
+  * To get the api key from https://gateway.seven.io
```
Comment (fchiyenda) [inline README.md:413]: This not clear please explain in detail how one can get the api key and also the url for the gateway

Comment (pamzey) [reply README.md:413]: I have revised the guide for getting the sms tell me if it makes sense

File: app/jobs/send_sms_job.rb (hunk 6 above)
Comment (pamzey) [reply app/jobs/send_sms_job.rb]: This has been fixed

File: config/initializers/sidekiq.rb (hunk 1 above)
Comment (pamzey) [reply config/initializers/sidekiq.rb:3]: This has been resolved

File: app/controllers/api/v1/immunization_reports_controller.rb (hunk 2 above)
Comment (pamzey) [reply app/controllers/api/v1/immunization_reports_controller.rb]: This was handled

File: config/initializers/load_application_yml.rb (hunk 4 above)
Comment (pamzey) [reply config/initializers/load_application_yml.rb]: This was handled

File: app/controllers/api/v1/immunization_reports_controller.rb (hunk 2 above)
Comment (pamzey) [reply app/controllers/api/v1/immunization_reports_controller.rb]: I have removed this file

**Review by petroskayange — APPROVED** [review]

**Review by fchiyenda — APPROVED** [review]

**Pull request merged**